patches/rust/0001-managarm-initial-port.patch

```diff
@@ -0,0 +1,616 @@
++    #[cfg(not(any(target_os = "dragonfly", target_os = "vxworks", target_os = "managarm")))]
+     #[cfg_attr(
+         any(
+             target_os = "linux",
+@@ -67,13 +67,13 @@
+ }
+ 
+ /// Returns the platform-specific value of errno
+-#[cfg(not(any(target_os = "dragonfly", target_os = "vxworks")))]
++#[cfg(not(any(target_os = "dragonfly", target_os = "vxworks", target_os = "managarm")))]
+ pub fn errno() -> i32 {
+     unsafe { (*errno_location()) as i32 }
+ }
+ 
+ /// Sets the platform-specific value of errno
+-#[cfg(all(not(target_os = "linux"), not(target_os = "dragonfly"), not(target_os = "vxworks")))] // needed for readdir and syscall!
++#[cfg(all(not(target_os = "linux"), not(target_os = "dragonfly"), not(target_os = "vxworks"), not(target_os = "managarm")))] // needed for readdir and syscall!
+ #[allow(dead_code)] // but not all target cfgs actually end up using it
+ pub fn set_errno(e: i32) {
+     unsafe { *errno_location() = e as c_int }
+@@ -111,6 +111,29 @@ pub fn set_errno(e: i32) {
+     }
+ }
+ 
++#[cfg(target_os = "managarm")]
++pub fn errno() -> i32 {
++    extern "C" {
++        #[thread_local]
++        static __mlibc_errno: c_int;
++    }
++
++    unsafe { __mlibc_errno as i32 }
++}
++
++#[cfg(target_os = "managarm")]
++#[allow(dead_code)]
++pub fn set_errno(e: i32) {
++    extern "C" {
++        #[thread_local]
++        static mut __mlibc_errno: c_int;
++    }
++
++    unsafe {
++        __mlibc_errno = e;
++    }
++}
++
+ /// Gets a detailed string description for the given error number.
+ pub fn error_string(errno: i32) -> String {
+     extern "C" {
+@@ -350,6 +373,11 @@ pub fn current_exe() -> io::Result<PathBuf> {
+     }
+ }
+ 
++#[cfg(target_os = "managarm")]
++pub fn current_exe() -> io::Result<PathBuf> {
++    unimplemented!()
++}
++
+ #[cfg(any(target_os = "macos", target_os = "ios"))]
+ pub fn current_exe() -> io::Result<PathBuf> {
+     extern "C" {
+diff --git a/library/std/src/sys/unix/thread.rs b/library/std/src/sys/unix/thread.rs
+index cda17eb4bd2..fa977f5755c 100644
+--- a/library/std/src/sys/unix/thread.rs
++++ b/library/std/src/sys/unix/thread.rs
+@@ -103,6 +103,13 @@ pub fn set_name(name: &CStr) {
+         }
+     }
+ 
++    #[cfg(target_os = "managarm")]
++    pub fn set_name(name: &CStr) {
++        unsafe {
++            libc::pthread_setname_np(libc::pthread_self(), name.as_ptr());
++        }
++    }
++
+     #[cfg(target_os = "netbsd")]
+     pub fn set_name(name: &CStr) {
+         use crate::ffi::CString;
+diff --git a/library/std/src/sys/unix/thread_local_dtor.rs b/library/std/src/sys/unix/thread_local_dtor.rs
+index c3275eb6f0e..8fe5def3ed7 100644
+--- a/library/std/src/sys/unix/thread_local_dtor.rs
++++ b/library/std/src/sys/unix/thread_local_dtor.rs
+@@ -15,7 +15,8 @@
+     target_os = "linux",
+     target_os = "fuchsia",
+     target_os = "redox",
+-    target_os = "emscripten"
++    target_os = "emscripten",
++    target_os = "managarm" // TODO: Is this right?
```

Comment:
Hm?

Reply:
Judge for yourself: https://github.com/rust-lang/rust/blob/07ef25984549bc33adbdc402e977655d8c9093a7/library/std/src/sys/unix/thread_local_dtor.rs#L7-L44

I think we don't need the branch, but I kept it just to be safe. What do you think?

Reply:
Yes, that should be fine.